### PR TITLE
8315935: [lworld] Apply flat renaming to C2 code

### PR DESCRIPTION
**Pull request by fparain**
Same renaming as in JDK-8315412 applied to C2 code.

Tested with Mach5, tier1

Fred

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8315935](https://bugs.openjdk.org/browse/JDK-8315935): [lworld] Apply flat renaming to C2 code (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/925/head:pull/925` \
`$ git checkout pull/925`

Update a local copy of the PR: \
`$ git checkout pull/925` \
`$ git pull https://git.openjdk.org/valhalla.git pull/925/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 925`

View PR using the GUI difftool: \
`$ git pr show -t 925`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/925.diff">https://git.openjdk.org/valhalla/pull/925.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/925#issuecomment-1711725793)

### CODE REVIEW

File: src/hotspot/share/opto/parse2.cpp [hunk 1]
```diff
@@ -278,7 +278,7 @@ void Parse::array_store(BasicType bt) {
         Node* casted_ary = ary;
         if (vk != nullptr && !stopped()) {
           // Element type is known, cast and store to flattened representation
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/parse2.cpp:280]:
```suggestion
          // Element type is known, cast and store to flat representation
```

File: src/hotspot/share/opto/subtypenode.cpp [hunk 2]
```diff
@@ -51,11 +51,11 @@ const Type* SubTypeCheckNode::sub(const Type* sub_t, const Type* super_t) const
   // Similar to logic in CmpPNode::sub()
   bool unrelated_classes = false;
   // Handle inline type arrays
-  if (subk->flatten_array() && superk->not_flatten_array()) {
+  if (subk->flat_array() && superk->not_flat_array()) {
     // The subtype is flattened in arrays and the supertype is not flattened in arrays. Must be unrelated.
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/subtypenode.cpp:55]:
```suggestion
    // The subtype is flat in arrays and the supertype is not flat in arrays. Must be unrelated.
```

File: src/hotspot/share/opto/type.cpp [hunk 3]
```diff
@@ -5198,16 +5198,16 @@ const Type *TypeAryPtr::xmeet_helper(const Type *t) const {
     if (meet_aryptr(ptr, elem, this, tap, res_klass, res_xk, res_flat, res_not_flat, res_not_null_free) == NOT_SUBTYPE) {
       instance_id = InstanceBot;
     } else if (this->is_flat() != tap->is_flat()) {
-      // Meeting flattened inline type array with non-flattened array. Adjust (field) offset accordingly.
+      // Meeting flat inline type array with non-flat array. Adjust (field) offset accordingly.
       if (tary->_flat) {
         // Result is flattened
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.cpp:5203]:
```suggestion
        // Result is flat
```

File: src/hotspot/share/opto/type.cpp [hunk 4]
```diff
@@ -5198,16 +5198,16 @@ const Type *TypeAryPtr::xmeet_helper(const Type *t) const {
...
         off = Offset(is_flat() ? offset() : tap->offset());
         field_off = is_flat() ? field_offset() : tap->field_offset();
       } else if (below_centerline(ptr)) {
         // Result is non-flattened
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.cpp:5207]:
```suggestion
        // Result is non-flat
```

**Review by TobiHartmann — APPROVED** [review]
Thanks a lot for taking care of this renaming, Fred! I found some remaining occurrences of "flattened", maybe there are more that need to be updated.

File: src/hotspot/share/opto/loopnode.hpp [hunk 5]
```diff
@@ -100,7 +100,7 @@ class LoopNode : public RegionNode {
   bool is_subword_loop() const { return _loop_flags & SubwordLoop; }
   bool is_loop_nest_inner_loop() const { return _loop_flags & LoopNestInnerLoop; }
   bool is_loop_nest_outer_loop() const { return _loop_flags & LoopNestLongOuterLoop; }
-  bool is_flattened_arrays() const { return _loop_flags & FlattenedArrays; }
+  bool is_flat_arrays() const { return _loop_flags & FlattenedArrays; }
```
[the remark this answers is not in the document]
Comment (fparain) [reply src/hotspot/share/opto/loopnode.hpp:103]: Yes, definitively.

File: src/hotspot/share/oops/flatArrayOop.hpp [hunk 6]
```diff
@@ -29,7 +29,8 @@
 #include "oops/klass.hpp"
 #include "runtime/handles.hpp"
 
-// A flatArrayOop is an array containing inline types (may include flatten embedded oop elements).
+// A flatArrayOop is an array containing flattened inline types (no indirection).
```
[the remark this answers is not in the document]
Comment (fparain) [reply src/hotspot/share/oops/flatArrayOop.hpp:32]: This is one aspect of the renaming that requires some bike-shading.
So far, I've applied the "flat" qualifier to the container, the field or the array, because there's a direct impact on the shape of the container and the way it is accessed.
But should we apply the "flat" qualifier to the value itself? The value is the value, a set of constants with a shape that doesn't change if it is stored in a flat field, a flat array or an standalone instance.
Both "flat inline type" and "flattened inline type" sound weird to me, so I don't really have a preference. Does one make more sense or is more informative than the other?

[the remark this answers is not in the document]
Comment (TobiHartmann) [reply src/hotspot/share/oops/flatArrayOop.hpp:32]: I don't have a strong opinion but maybe we should simply rephrase this to:
`A flatArrayOop is a flat array containing inline types (no indirection).`

Or

`A flatArrayOop points to a flat array containing inline types (no indirection).`

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. Thanks!

**Comment by fparain** [comment]
Thank you for your review and feedback.
